### PR TITLE
Add FW Filter Slider Support

### PR DESCRIPTION
**Pull request by spatzengr**
Added firmware side filter slider support.

**BIG CREDIT** to @haslinghuis as **he MADE THIS WORK**!!  Without his help, this would have not got done; under this PR at least.  

In the current Configurator Master, the firmware side filter sliders for API 1.44 and greater are not yet supported.  Calculations you see in the Configurator for the lowpass filter are based on legacy Configurator sliders functions (in current Master).  This PR adds TRUE support for BF 4.3 firmware-based filter sliders.  

It also fixed the bug with the PIDs grid in API <= 1.43 (BF 4.2 and lower) whereas when you type in the PID gain manually - disable the sliders - it is reset back to the previous value.  That should not happen for BF 4.2 and lower.

### CODE REVIEW

File: src/js/TuningSliders.js [hunk 1]
```diff
@@ -414,61 +449,97 @@ TuningSliders.calculateNewPids = function() {
...
+
+        Promise.resolve(true)
+        .then(() => { return MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS)); })
+        .then(() => { return MSP.send_message(MSPCodes.MSP_FILTER_CONFIG); });
+
+        const delayInMilliseconds = 250; //0.25 seconds
+        setTimeout(function() {
+            TuningSliders.updateLowpassValues();
+        }, delayInMilliseconds);
+    }
```
Comment (chmelevskij) [inline src/js/TuningSliders.js:506]: There is not point of doing `Promise.resolve(true)` here,`MSP.promise` already returns a promise. Also why that timeout? Can't we update low pass values after the last promise resolves?

Something like?
```suggestion
        const delayInMilliseconds = 250; //0.25 seconds
        MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS))
        .then(() => MSP.send_message(MSPCodes.MSP_FILTER_CONFIG))
        .then(() => TuningSliders.updateLowpassValues())
    }
```

File: src/js/TuningSliders.js [hunk 2]
```diff
@@ -414,61 +449,97 @@ TuningSliders.calculateNewPids = function() {
...
+        FC.TUNING_SLIDERS.slider_dterm_filter_multiplier = this.sliderDTermFilterMultiplier * 100;
+
+        Promise.resolve(true)
+        .then(() => { return MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS)); })
+        .then(() => { return MSP.send_message(MSPCodes.MSP_FILTER_CONFIG); });
+
+        const delayInMilliseconds = 250; //0.25 seconds
+        setTimeout(function() {
+            TuningSliders.updateLowpassValues();
+        }, delayInMilliseconds);
```
Comment (chmelevskij) [inline src/js/TuningSliders.js:527]: Same here, `Promise.resolve` is unnecessary, and would be nice to check if we can avoid `setTimeout`

File: src/js/TuningSliders.js [hunk 3]
```diff
@@ -181,33 +185,39 @@ TuningSliders.resetPidSliders = function() {
...
         this.initDTermFilterSliderPosition();
     }
 
     this.calculateNewDTermFilters();
-    this.updateFilterSlidersDisplay();
+
+    const delayInMilliseconds = 300; //0.3 seconds
+    setTimeout(function() {
+        TuningSliders.updateFilterSlidersDisplay();
+    }, delayInMilliseconds);
```
Comment (chmelevskij) [inline src/js/TuningSliders.js:220]: Why `setTimeout`? In general it leads to possible memory leaks, E.G. you do something, and then navigate away. Is there a better signal? Like for example sending a message and waiting for promise to resolve?

File: src/js/TuningSliders.js (hunk 2 above)
Comment (spatzengr) [reply src/js/TuningSliders.js:527]: @chmelevskij , unfortunately without the Timeout function, I'm getting this behavior.  It takes somewhere between 1ms to 2.5ms for the board to calc the new PID values and returns them to the configurator.  So we need to wait for that to be done.  I'm thinking the ".then" function is only waiting till the Configurator does the line above, which doesn't take into account the delay of the FC and USB interface.  If you know of a better way than Timeout, I'm all ears.

2IsE8WVtmj

File: src/js/TuningSliders.js (hunk 1 above)
Comment (spatzengr) [reply src/js/TuningSliders.js:506]: @chmelevskij , as outlined below - which is a little simpler to demonstrate the lag - if I try to implement with the filter sliders, since the calcs take 1 to 3.5ms to complete (D-term filter takes the longest), it will shut off the slider if you call `TuningSliders.updateLowpassValues()` and `TuningSliders.updateFilterSlidersDisplay()` in succession without a Timeout.

iPJu2nEYd3

Comment (chmelevskij) [reply src/js/TuningSliders.js:506]: `send_message`  doesn't return the promise though, so even though you use `then` it resolves straight away. Can you use `.promise` for that as well?

You don't need `Promise.resolve(true)` though ,with or without the timeout

Comment (spatzengr) [reply src/js/TuningSliders.js:506]: @chmelevskij , the promise seems to work well to avoid the time-out to call one function after MPS, thank you!
However, calling two functions still seems to glitch out.  Any suggestions?

image

fretwUlIp6

Comment (chmelevskij) [reply src/js/TuningSliders.js:506]: Do you get the same if you put both functions in the same promise resolve?

something like:
```
MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS))
  .then(() => MSP.promise(MSPCodes.MSP_FILTER_CONFIG)
  .then(() => {
    TuningSliders.updateLowpassValues();
    TuningSliders.updateFiltersSlidersDisplay();
  })
```

Comment (spatzengr) [reply src/js/TuningSliders.js]: @chmelevskij that worked!!  Thanks again.  Now the only glitch is you have to press Enables Sliders twice if you turn off a lowpass or adjust a value manually and Save.

Updated code is committed.  
8NZKYj8XqU

File: src/js/TuningSliders.js [hunk 4]
```diff
@@ -380,95 +419,116 @@ TuningSliders.legacyCalculateNewPids = function() {
...
+    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
+        FC.TUNING_SLIDERS.slider_gyro_filter = this.sliderGyroFilter;
+        //rounds slider values to nearies multiple of 5 and passes to the FW
+        FC.TUNING_SLIDERS.slider_gyro_filter_multiplier = Math.round((this.sliderGyroFilterMultiplier * 100)/5)*5;
+
+        MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS))
+        .then(() => MSP.promise(MSPCodes.MSP_FILTER_CONFIG))
+        .then(() => {
+          this.updateLowpassValues();
+          this.updateFiltersSlidersDisplay();
```
Comment (haslinghuis) [inline src/js/TuningSliders.js:493]:
```suggestion
          this.updateFilterSlidersDisplay();
```

File: src/js/TuningSliders.js [hunk 5]
```diff
@@ -380,95 +419,116 @@ TuningSliders.legacyCalculateNewPids = function() {
...
+    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
+        FC.TUNING_SLIDERS.slider_dterm_filter = this.sliderDTermFilter;
+        //rounds slider values to nearies multiple of 5 and passes to the FW
+        FC.TUNING_SLIDERS.slider_dterm_filter_multiplier = Math.round((this.sliderDTermFilterMultiplier * 100)/5)*5;
+
+        MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS))
+        .then(() => MSP.promise(MSPCodes.MSP_FILTER_CONFIG))
+        .then(() => {
+            this.updateLowpassValues();
+            this.updateFiltersSlidersDisplay();
```
Comment (haslinghuis) [inline src/js/TuningSliders.js:515]:
```suggestion
            this.updateFilterSlidersDisplay();
```

File: src/js/TuningSliders.js [hunk 6]
```diff
@@ -207,7 +210,6 @@ TuningSliders.resetDTermFilterSlider = function() {
     }
 
     this.calculateNewDTermFilters();
-    this.updateFilterSlidersDisplay();
```
Comment (haslinghuis) [inline src/js/TuningSliders.js:210]: Maybe this one should stay

**Review by haslinghuis — CHANGES_REQUESTED** [review]

File: src/js/TuningSliders.js (hunk 6 above)
Comment (haslinghuis) [reply src/js/TuningSliders.js:210]: With above changes we don't have to press the enable the DTerm slider twice after disabling Dterm settings in legacy mode.

File: src/js/TuningSliders.js [hunk 7]
```diff
@@ -380,95 +419,116 @@ TuningSliders.legacyCalculateNewPids = function() {
...
+        this.calculateLegacyGyroFilters();
+
+    }
+
+    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
+        FC.TUNING_SLIDERS.slider_gyro_filter = this.sliderGyroFilter;
+        //rounds slider values to nearies multiple of 5 and passes to the FW
+        FC.TUNING_SLIDERS.slider_gyro_filter_multiplier = Math.round((this.sliderGyroFilterMultiplier * 100)/5)*5;
+
+        MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS))
```
Comment (haslinghuis) [inline src/js/TuningSliders.js:489]: Maybe make this part a separate function, its duplicated in 511

**Comment by haslinghuis** [comment]
Gyro filters have same behavior as dterm before (need double enable), Dterm is fixed now (legacy)

**Comment by haslinghuis** [comment]
@spatzengr 

Debugging and testing your branch with some changes and it seems working on `legacy` and `firmware` as far I have tested without need for timeouts. On `legacy` I think we don't need the check for `cachedGyroSliderValues` and `cachedDTermSliderValues` anymore. So far I can't tell the difference as it's called after clicking `enable sliders` anyway.
```
TuningSliders.resetGyroFilterSlider = function() {

    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
        if (!this.cachedGyroSliderValues) {
            this.sliderGyroFilterMultiplier = 1;
        }
        this.initGyroFilterSliderPosition();
    } else {
        this.sliderGyroFilterMultiplier = 1;
        $('#sliderGyroFilterMultiplier').val(this.downscaleSliderValue(this.sliderGyroFilterMultiplier));
    }

    this.calculateNewGyroFilters();
    TuningSliders.updateFilterSlidersDisplay();
};

TuningSliders.resetDTermFilterSlider = function() {

    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
        if (!this.cachedDTermSliderValues) {
            this.sliderDTermFilterMultiplier = 1;
        }
        this.initDTermFilterSliderPosition();
    } else {
        this.sliderDTermFilterMultiplier = 1;
        $('#sliderDTermFilterMultiplier').val(this.downscaleSliderValue(this.sliderDTermFilterMultiplier));
    }

    this.calculateNewDTermFilters();
    TuningSliders.updateFilterSlidersDisplay();
};
```

On the firmware side remaining after enabling the normal pid filter you have to `refresh` after `enable sliders`. Adding
```
                TuningSliders.updatePidSlidersDisplay();
```
on line 1931 (pid_tuning.js) should solve that too.

**Comment by spatzengr** [comment]
@haslinghuis 

With the above, I'm still getting the behavior of with the sliders set to a value other than 1.0, if you turn off a Dynamic Lowpass and save, you have to press the Enable button twice. 

I also have noticed that if the Gyro Slider is set to 1.15, it won't allow it to re-enable.  I'm presuming it is a coincidence with the defaults for Gyro filters, 1.15 and the `Math.floor` function used in `updateFiltersliderDisplay()`

Uf40XOJHqp

**Comment by haslinghuis** [comment]
@spatzengr it's weird. Already tested with the exact steps the video is showing but on Linux it's just working with one click.

**Comment by haslinghuis** [comment]
@spatzengr Got it working on my side: please try the changes in the diff: [fix_2429_diff.txt](https://github.com/betaflight/betaflight-configurator/files/6118174/fix_2429_diff.txt)

**Comment by haslinghuis** [comment]
[fix_2429.diff.txt](https://github.com/betaflight/betaflight-configurator/files/6118666/fix_2429.diff.txt)

**Comment by haslinghuis** [comment]
@spatzengr this is dealing with odd `input` or `slider` values in firmware. The `abs` method is one value off sometimes but it's more foregiving... you can see the values changing when using sliders and it's validated all the time.
```
TuningSliders.updateFilterSlidersDisplay = function() {
    // check if filters changed manually by comapring current value and those based on slider position
    const WARNING_FILTER_HIGH_GAIN = 1.4;
    const WARNING_FILTER_LOW_GAIN = 0.7;

    this.GyroSliderUnavailable = false;
    this.DTermSliderUnavailable = false;

    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
        console.log('Slider parameters: ', this.sliderGyroFilter, this.sliderDTermFilter);
        if (this.sliderGyroFilter === 0) {
            this.GyroSliderUnavailable = true;
        }
        if (this.sliderDtermFilter === 0) {
            this.DTermSliderUnavailable = true;
        }
    }

    console.log('------------------------------------------------------------------------------');
    console.log('gyro_low_pass_dyn_min_hz', parseInt($('.pid_filter input[name="gyroLowpassDynMinFrequency"]').val()), Math.round(this.FILTER_DEFAULT.gyro_lowpass_dyn_min_hz * this.sliderGyroFilterMultiplier));
    console.log('gyro_lowpass_dyn_max_hz', parseInt($('.pid_filter input[name="gyroLowpassDynMaxFrequency"]').val()), Math.floor(this.FILTER_DEFAULT.gyro_lowpass_dyn_max_hz * this.sliderGyroFilterMultiplier));
    console.log('gyro_lowpass_type', parseInt($('.pid_filter select[name="gyroLowpassDynType"]').val()), this.FILTER_DEFAULT.gyro_lowpass_type);
    console.log('gyro_lowpass2_hz', parseInt($('.pid_filter input[name="gyroLowpass2Frequency"]').val()), Math.floor(this.FILTER_DEFAULT.gyro_lowpass2_hz * this.sliderGyroFilterMultiplier));
    console.log('gyro_lowpass2_type', parseInt($('.pid_filter select[name="gyroLowpass2Type"]').val()), this.FILTER_DEFAULT.gyro_lowpass2_type);

    console.log(Math.abs(parseInt($('.pid_filter input[name="gyroLowpassDynMinFrequency"]').val()) - (this.FILTER_DEFAULT.gyro_lowpass_dyn_min_hz * this.sliderGyroFilterMultiplier)));
    console.log(Math.abs(parseInt($('.pid_filter input[name="gyroLowpassDynMaxFrequency"]').val()) - (this.FILTER_DEFAULT.gyro_lowpass_dyn_max_hz * this.sliderGyroFilterMultiplier)));
    console.log(Math.abs(parseInt($('.pid_filter select[name="gyroLowpassDynType"]').val()) - this.FILTER_DEFAULT.gyro_lowpass_type));
    console.log(Math.abs(parseInt($('.pid_filter input[name="gyroLowpass2Frequency"]').val()) - (this.FILTER_DEFAULT.gyro_lowpass2_hz * this.sliderGyroFilterMultiplier)));
    console.log(Math.abs(parseInt($('.pid_filter select[name="gyroLowpass2Type"]').val()) - this.FILTER_DEFAULT.gyro_lowpass2_type));


    if (Math.abs(parseInt($('.pid_filter input[name="gyroLowpassDynMinFrequency"]').val()) - (this.FILTER_DEFAULT.gyro_lowpass_dyn_min_hz * this.sliderGyroFilterMultiplier)) > 1 ||
        Math.abs(parseInt($('.pid_filter input[name="gyroLowpassDynMaxFrequency"]').val()) - (this.FILTER_DEFAULT.gyro_lowpass_dyn_max_hz * this.sliderGyroFilterMultiplier)) > 1 ||
        Math.abs(parseInt($('.pid_filter select[name="gyroLowpassDynType"]').val()) - this.FILTER_DEFAULT.gyro_lowpass_type) > 1 ||
        Math.abs(parseInt($('.pid_filter input[name="gyroLowpass2Frequency"]').val()) - (this.FILTER_DEFAULT.gyro_lowpass2_hz * this.sliderGyroFilterMultiplier)) > 1 ||
        Math.abs(parseInt($('.pid_filter select[name="gyroLowpass2Type"]').val()) - this.FILTER_DEFAULT.gyro_lowpass2_type) > 1) {
        this.GyroSliderUnavailable = true;
        if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
            console.log('Disabling gyro filter slider because values do not match with firmware');
            this.sliderGyroFilter = 0;
        }
    } else {
        this.GyroSliderUnavailable = false;
        this.cachedGyroSliderValues = true;
        this.sliderGyroFilter = 1;
    }

    console.log('dtermLowpassDynMinFrequency', parseInt($('.pid_filter input[name="dtermLowpassDynMinFrequency"]').val()), Math.floor(this.FILTER_DEFAULT.dterm_lowpass_dyn_min_hz * this.sliderDTermFilterMultiplier));
    console.log('dtermLowpassDynMaxFrequency', parseInt($('.pid_filter input[name="dtermLowpassDynMaxFrequency"]').val()), Math.floor(this.FILTER_DEFAULT.dterm_lowpass_dyn_max_hz * this.sliderDTermFilterMultiplier));
    console.log('dtermLowpassDynType', parseInt($('.pid_filter select[name="dtermLowpassDynType"]').val()), this.FILTER_DEFAULT.dterm_lowpass_type);
    console.log('dtermLowpass2Frequency', parseInt($('.pid_filter input[name="dtermLowpass2Frequency"]').val()), Math.floor(this.FILTER_DEFAULT.dterm_lowpass2_hz * this.sliderDTermFilterMultiplier));
    console.log('dtermLowpass2Type', parseInt($('.pid_filter select[name="dtermLowpass2Type"]').val()), this.FILTER_DEFAULT.dterm_lowpass2_type);
    console.log('------------------------------------------------------------------------------');

    console.log(Math.abs(parseInt($('.pid_filter input[name="dtermLowpassDynMinFrequency"]').val()) - (this.FILTER_DEFAULT.dterm_lowpass_dyn_min_hz * this.sliderDTermFilterMultiplier)));
    console.log(Math.abs(parseInt($('.pid_filter input[name="dtermLowpassDynMaxFrequency"]').val()) - (this.FILTER_DEFAULT.dterm_lowpass_dyn_max_hz * this.sliderDTermFilterMultiplier)));
    console.log(Math.abs(parseInt($('.pid_filter select[name="dtermLowpassDynType"]').val()) - this.FILTER_DEFAULT.dterm_lowpass_type));
    console.log(Math.abs(parseInt($('.pid_filter input[name="dtermLowpass2Frequency"]').val()) - (this.FILTER_DEFAULT.dterm_lowpass2_hz * this.sliderDTermFilterMultiplier)));
    console.log(Math.abs(parseInt($('.pid_filter select[name="dtermLowpass2Type"]').val()) - this.FILTER_DEFAULT.dterm_lowpass2_type));

    if (Math.abs(parseInt($('.pid_filter input[name="dtermLowpassDynMinFrequency"]').val()) - (this.FILTER_DEFAULT.dterm_lowpass_dyn_min_hz * this.sliderDTermFilterMultiplier)) > 1 ||
        Math.abs(parseInt($('.pid_filter input[name="dtermLowpassDynMaxFrequency"]').val()) - (this.FILTER_DEFAULT.dterm_lowpass_dyn_max_hz * this.sliderDTermFilterMultiplier)) > 1 ||
        Math.abs(parseInt($('.pid_filter select[name="dtermLowpassDynType"]').val()) - this.FILTER_DEFAULT.dterm_lowpass_type) > 1 ||
        Math.abs(parseInt($('.pid_filter input[name="dtermLowpass2Frequency"]').val()) - (this.FILTER_DEFAULT.dterm_lowpass2_hz * this.sliderDTermFilterMultiplier)) > 1 ||
        Math.abs(parseInt($('.pid_filter select[name="dtermLowpass2Type"]').val()) - this.FILTER_DEFAULT.dterm_lowpass2_type) > 1) {
        this.DTermSliderUnavailable = true;
        if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
            console.log('Disable DTerm slider because values do not match with firmware');
            this.sliderDTermFilter = 0;
        }
    } else {
        this.DTermSliderUnavailable = false;
        this.cachedDTermSliderValues = true;
        this.sliderDTermFilter = 1;
    }

    if (this.GyroSliderUnavailable) {
        $('.tuningFilterSliders .sliderLabels tr:nth-child(2)').hide();
    } else {
        $('.tuningFilterSliders .sliderLabels tr:nth-child(2)').show();
    }

    if (this.DTermSliderUnavailable) {
        $('.tuningFilterSliders .sliderLabels tr:last-child').hide();
    } else {
        $('.tuningFilterSliders .sliderLabels tr:last-child').show();
    }

    $('.tuningFilterSliders').toggle(!(this.GyroSliderUnavailable && this.DTermSliderUnavailable));
    $('.subtab-filter .slidersDisabled').toggle(this.GyroSliderUnavailable || this.DTermSliderUnavailable);
    $('.subtab-filter .nonExpertModeSlidersNote').toggle((!this.GyroSliderUnavailable || !this.DTermSliderUnavailable) && !this.expertMode);
    $('.subtab-filter .slidersWarning').toggle(((this.sliderGyroFilterMultiplier >= WARNING_FILTER_HIGH_GAIN ||
                                                    this.sliderGyroFilterMultiplier <= WARNING_FILTER_LOW_GAIN) && !this.GyroSliderUnavailable) ||
                                                    ((this.sliderDTermFilterMultiplier >= WARNING_FILTER_HIGH_GAIN ||
                                                     this.sliderDTermFilterMultiplier <= WARNING_FILTER_LOW_GAIN) && !this.DTermSliderUnavailable));
};
```

**Comment by haslinghuis** [comment]
[diff.txt](https://github.com/betaflight/betaflight-configurator/files/6133944/diff.txt)

**Comment by haslinghuis** [comment]
Fixes firmware calculation in UI and also fixes a bug in legacy calculation of the `dterm` sliders in legacy - changed to `round()`

[diff.txt](https://github.com/betaflight/betaflight-configurator/files/6134097/diff.txt)

**Review by haslinghuis — APPROVED** [review]
Tested - also fixes legacy slider calculation (as observed in firmware)

**Comment by nangel0** [comment]
> Tested - also fixes legacy slider calculation (as observed in firmware)

This fixes you talk about is the Gyro lowpass buttons disabling not being saved? I have observed this behavior (when i try to enable only a single gyro notch I cant disable Gyro low pass filters, and slider, when I it save it reverts everuthing to SLiders at 1)

Not sure if this is the correct place, i didnt what to open uneccessary issues.

**Review by McGiverGim — DISMISSED** [review]

File: src/js/TuningSliders.js [hunk 8]
```diff
@@ -133,10 +135,11 @@ TuningSliders.initPidSlidersPosition = function() {
 };
 
 TuningSliders.initGyroFilterSliderPosition = function() {
-    if (semver.lte(FC.CONFIG.apiVersion, API_VERSION_1_43)) {
+    if (semver.lt(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
```
Comment (mikeller) [inline src/js/TuningSliders.js:138]: What is the reason for these changes? Pretty much all existing checks in this file are `semver.lte(FC.CONFIG.apiVersion, API_VERSION_1_43)`, so changing this here just adds inconsistency and makes this harder to maintain.

File: src/js/TuningSliders.js [hunk 9]
```diff
@@ -177,37 +181,74 @@ TuningSliders.resetPidSliders = function() {
     }
 
     this.calculateNewPids();
-    this.updatePidSlidersDisplay();
 };
 
 TuningSliders.resetGyroFilterSlider = function() {
-    if (!this.cachedGyroSliderValues) {
+    if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
```
Comment (mikeller) [inline src/js/TuningSliders.js:187]: Same as above - there are at least 4 ways that each version check can be formulated (using `lt`, `lte`, `gt`, `gte`). Using a mix of all of these to check the same condition in the same file will make this hard to maintain.

File: src/js/tabs/pid_tuning.js [hunk 10]
```diff
@@ -995,19 +995,19 @@ TABS.pid_tuning.initialize = function (callback) {
...
         if (semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)) {
             FC.TUNING_SLIDERS.slider_pids_mode = parseInt($('#sliderPidsModeSelect').val());
-            FC.TUNING_SLIDERS.slider_master_multiplier = TuningSliders.sliderMasterMultiplier * 100;
-            FC.TUNING_SLIDERS.slider_roll_pitch_ratio = TuningSliders.sliderRollPitchRatio * 100;
-            FC.TUNING_SLIDERS.slider_i_gain = TuningSliders.sliderIGain * 100;
-            FC.TUNING_SLIDERS.slider_pd_ratio = TuningSliders.sliderPDRatio * 100;
-            FC.TUNING_SLIDERS.slider_pd_gain = TuningSliders.sliderPDGain * 100;
-            FC.TUNING_SLIDERS.slider_dmin_ratio = TuningSliders.sliderDMinRatio * 100;
-            FC.TUNING_SLIDERS.slider_ff_gain = TuningSliders.sliderFFGain * 100;
+            FC.TUNING_SLIDERS.slider_master_multiplier = Math.round((TuningSliders.sliderMasterMultiplier * 100)/5)*5;
```
Comment (mikeller) [inline src/js/tabs/pid_tuning.js:998]: Please put spaces around operators. Also, `...* 100 / 5` can be simplified to `...* 20`. ;-)

File: src/js/tabs/pid_tuning.js [hunk 11]
```diff
@@ -995,19 +995,19 @@ TABS.pid_tuning.initialize = function (callback) {
...
+            FC.TUNING_SLIDERS.slider_roll_pitch_ratio = Math.round((TuningSliders.sliderRollPitchRatio * 100)/5)*5;
+            FC.TUNING_SLIDERS.slider_i_gain = Math.round((TuningSliders.sliderIGain * 100)/5)*5;
+            FC.TUNING_SLIDERS.slider_pd_ratio = Math.round((TuningSliders.sliderPDRatio * 100)/5)*5;
+            FC.TUNING_SLIDERS.slider_pd_gain = Math.round((TuningSliders.sliderPDGain * 100)/5)*5;
+            FC.TUNING_SLIDERS.slider_dmin_ratio = Math.round((TuningSliders.sliderDMinRatio * 100)/5)*5;
+            FC.TUNING_SLIDERS.slider_ff_gain = Math.round((TuningSliders.sliderFFGain * 100)/5)*5;
 
-            FC.TUNING_SLIDERS.slider_dterm_filter = TuningSliders.sliderDTermFilter ? 1 : 0;
-            FC.TUNING_SLIDERS.slider_dterm_filter_multiplier = TuningSliders.sliderDTermFilterMultiplier * 100;
+            FC.TUNING_SLIDERS.slider_dterm_filter = TuningSliders.sliderDTermFilter;
```
Comment (mikeller) [inline src/js/tabs/pid_tuning.js:1006]: Do we want to allow arbitrary values in here?

File: src/js/tabs/pid_tuning.js (hunk 10 above)
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js:998]: LOL... me feel like a dumb-dumb.  Will change.  Was not thinking about dividing optimization.  My bad.

File: src/js/TuningSliders.js (hunk 8 above)
Comment (haslinghuis) [reply src/js/TuningSliders.js:138]: No problem, I wanted to use `API_VERSION_1_44` for consistency.

Comment (spatzengr) [reply src/js/TuningSliders.js:138]: I changed everything in the file to be .le for 1_44.

File: src/js/TuningSliders.js (hunk 9 above)
Comment (spatzengr) [reply src/js/TuningSliders.js:187]: made all the same

File: src/js/tabs/pid_tuning.js (hunk 11 above)
Comment (spatzengr) [reply src/js/tabs/pid_tuning.js:1006]: rounded with x 20.

File: src/js/TuningSliders.js (hunk 8 above)
Comment (haslinghuis) [reply src/js/TuningSliders.js:138]: To me this is much more readable. @mikeller ?

Comment (haslinghuis) [reply src/js/TuningSliders.js:138]: To resolve we could
```
Add line 31: `firmwareSliders: false,`
and line 48: `this.firmwareSliders = semver.gte(FC.CONFIG.apiVersion, API_VERSION_1_44)
```
And use this variable instead of the comparision

Comment (spatzengr) [reply src/js/TuningSliders.js:138]: @haslinghuis , I think we're good.  I just changed all to .lt(44) so they are consistent within the files.  I think 3 were still the other way.  Easy peasy lemon squeezy.

File: src/js/TuningSliders.js [hunk 12]
```diff
@@ -380,95 +457,111 @@ TuningSliders.legacyCalculateNewPids = function() {
...
-        FC.TUNING_SLIDERS.slider_gyro_filter_multiplier = TuningSliders.sliderGyroFilterMultiplier * 100;
-
-        Promise.resolve(true)
-        .then(() => { return MSP.promise(MSPCodes.MSP_SET_TUNING_SLIDERS, mspHelper.crunch(MSPCodes.MSP_SET_TUNING_SLIDERS)); })
-        .then(() => { return MSP.send_message(MSPCodes.MSP_PID); })
-        .then(() => { return MSP.send_message(MSPCodes.MSP_PID_ADVANCED); });
+        //rounds slider values to nearies multiple of 5 and passes to the FW. Avoid divide by (var x 100)/5 = 20
+        FC.TUNING_SLIDERS.slider_master_multiplier = Math.round(this.sliderMasterMultiplier * 20)*5;
+        FC.TUNING_SLIDERS.slider_roll_pitch_ratio = Math.round(this.sliderRollPitchRatio * 20)*5;
+        FC.TUNING_SLIDERS.slider_i_gain = Math.round(this.sliderIGain * 20)*5;
```
Comment (haslinghuis) [inline src/js/TuningSliders.js:477]: Please give operators  some  space

File: src/js/tabs/pid_tuning.js [hunk 13]
```diff
@@ -995,19 +995,20 @@ TABS.pid_tuning.initialize = function (callback) {
...
             FC.TUNING_SLIDERS.slider_pids_mode = parseInt($('#sliderPidsModeSelect').val());
-            FC.TUNING_SLIDERS.slider_master_multiplier = TuningSliders.sliderMasterMultiplier * 100;
-            FC.TUNING_SLIDERS.slider_roll_pitch_ratio = TuningSliders.sliderRollPitchRatio * 100;
-            FC.TUNING_SLIDERS.slider_i_gain = TuningSliders.sliderIGain * 100;
-            FC.TUNING_SLIDERS.slider_pd_ratio = TuningSliders.sliderPDRatio * 100;
-            FC.TUNING_SLIDERS.slider_pd_gain = TuningSliders.sliderPDGain * 100;
-            FC.TUNING_SLIDERS.slider_dmin_ratio = TuningSliders.sliderDMinRatio * 100;
-            FC.TUNING_SLIDERS.slider_ff_gain = TuningSliders.sliderFFGain * 100;
+            //rounds slider values to nearies multiple of 5 and passes to the FW. Avoid divide by (var x 100)/5 = 20
+            FC.TUNING_SLIDERS.slider_master_multiplier = Math.round(TuningSliders.sliderMasterMultiplier * 20)*5;
```
Comment (haslinghuis) [inline src/js/tabs/pid_tuning.js:999]: Please give operators some 👾 
```suggestion
            FC.TUNING_SLIDERS.slider_master_multiplier = Math.round(TuningSliders.sliderMasterMultiplier * 20) * 5;
```

**Review by haslinghuis — CHANGES_REQUESTED** [review]

**Comment by spatzengr** [comment]
@haslinghuis , I have been noticing that when one slider is disabled, it was calling `this.updateFilterSlidersDisplay();` (as intended with an || statement) which was causing the sliders to "blink" if you moved the other slider.  So with this operator spacing cleanup, I changed to only run if coming from clicking the "Enable Slider" with a boolean.

**Review by mikeller — APPROVED** [review]
